compression-converter/data-converter.go

```diff
@@ -0,0 +1,187 @@
+package compressionconverter
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	commonpb "go.temporal.io/api/common/v1"
+
+	"go.temporal.io/sdk/converter"
+
+	"github.com/pierrec/lz4"
+)
+
+const (
+	// MetadataCompressionAlgorithmKey is "compression-algorithm"
+	MetadataCompressionAlgorithmKey = "compression-algorithm"
+
+	// MetadataCompressionAlgorithm is "lz4"
+	MetadataCompressionAlgorithm = "lz4"
+
+	// MetadataEncodingCompressed is "binary/compressed"
+	MetadataEncodingCompressed = "binary/compressed"
+)
+
+// CompressionConverter implements a DataConverter using LZ4.
+type CompressionConverter struct {
+	dataConverter converter.DataConverter
+}
+
+// NewCompressionConverter creates a new instance of CompressionConvertter wrapping a DataConverter
+func NewCompressionConverter(dataConverter converter.DataConverter) *CompressionConverter {
+	return &CompressionConverter{
+		dataConverter: dataConverter,
+	}
+}
+
+// ToPayloads converts a list of values.
+func (dc *CompressionConverter) ToPayloads(values ...interface{}) (*commonpb.Payloads, error) {
+	result := &commonpb.Payloads{}
+
+	for i, value := range values {
```

Comment:
`Payloads` doesn't have metadata so we wouldn't be able to record that it was compressed, or with which algorithm. That might be ok, but feels a little brittle to me. I agree it would be preferable for performance.

Reply:
You can compress all the payloads into a single Payload that contains the metadata.